eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelType.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2021 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.channel
+
+import fr.acinq.eclair.FeatureSupport.Optional
+import fr.acinq.eclair.Features
+import fr.acinq.eclair.Features.{AnchorOutputs, StaticRemoteKey}
+import fr.acinq.eclair.transactions.Transactions.{AnchorOutputsCommitmentFormat, CommitmentFormat, DefaultCommitmentFormat}
+
+/**
+ * Created by t-bast on 24/06/2021.
+ */
+
+/**
+ * A channel type is a specific combination of Bolt 9 features, defined in the RFC (Bolt 2).
+ */
+case class ChannelType(features: Features) {
```

Comment:
How about something with more types?

```scala
/**
 * A channel type is a specific combination of Bolt 9 features, defined in the RFC (Bolt 2).
 */
sealed trait ChannelType {
  def features: Features

  def commitmentFormat: CommitmentFormat = {
    if (features.hasFeature(AnchorOutputs)) {
      AnchorOutputsCommitmentFormat
    } else {
      DefaultCommitmentFormat
    }
  }

  /**
   * True if our main output in the remote commitment is directly sent (without any delay) to one of our wallet addresses.
   */
  def paysDirectlyToWallet: Boolean = {
    features.hasFeature(Features.StaticRemoteKey) && !features.hasFeature(Features.AnchorOutputs)
  }
}

object ChannelTypes {

  // @formatter:off
  case object Standard extends ChannelType { override def features: Features = Features.empty }
  case object StaticRemoteKey extends ChannelType { override def features: Features = Features(Features.StaticRemoteKey -> Optional) }
  case object AnchorOutputs extends ChannelType { override def features: Features = Features(Features.StaticRemoteKey -> Optional, Features.AnchorOutputs -> Optional) }
  // @formatter:on

  /**
   * Pick the channel type that should be applied based on features alone (in case our peer doesn't support explicit channel type negotiation).
   */
  def pickChannelType(localFeatures: Features, remoteFeatures: Features): ChannelType = {
    if (Features.canUseFeature(localFeatures, remoteFeatures, Features.AnchorOutputs)) {
      AnchorOutputs
    } else if (Features.canUseFeature(localFeatures, remoteFeatures, Features.StaticRemoteKey)) {
      StaticRemoteKey
    } else {
      Standard
    }
  }

}
```

Reply:
Good idea, I like it!

Reply:
It woks very well when features are exclusive (either A or B or C) but probably not that well with features that can be activate with any other combination of features (like `option_upfront_shutdown_script` or `peer_backup`) ?

Reply:
True, that's why I didn't do it at first, to keep the flexibility of orthogonal features without having to list the explosion of different channel types explicitly (as this comment shows https://github.com/lightningnetwork/lightning-rfc/pull/880#discussion_r655962589). Having a `sealed trait` is nice when we're pattern matching on the channel type, but I think it's not really necessary here, so I'll keep it like it was for now.

Note that I don't think `option_upfront_shutdown_script` nor `peer_backup` should be in the channel features, they aren't features that affect the channel's behavior (and we don't need them here to implement their behavior).

Reply:
Those are two different topics though:
1) types vs values
2) exclusive vs additive

More precisely on 2): do we define a limited number of types that are a combination of certain features, or should we define all types that are a combination of all supported features?

If the answer to 2) is _no_, then I believe types are a better implementation than values. The codec would be a mapping between the whole feature bit and an `Unsupported` type would be used as a catch all.

Now, if the answer to 2) is _yes_, then neither types nor values are a good implementation, because we would have to define a lot of them. But I believe this is not the spirit of the spec, otherwise we would match two bitvectors at channel opening, instead of two _arrays_ of bitvectors. In this case, the codec would operate at the bit level.

Reply:
I remove `ChannelType` from this PR entirely, we now simply store the channel bolt 9 features (from which we can later extract the channel type, once spec-ed).